### PR TITLE
[vcpkg registries] Add git registries

### PR DESCRIPTION
**Pull request by strega-nil**
Add git registries support to the registries module of vcpkg.

Depends on #15583

### CODE REVIEW

File: toolsrc/src/vcpkg/configuration.cpp [hunk 1]
```diff
@@ -28,17 +30,24 @@ namespace vcpkg
...
             registries_feature_flags_warning = true;
             regs.clear();
         }
 
+        if (!r.errors().empty())
+        {
+            System::print2(System::Color::error, "Errors occurred while parsing configuration.\n");
+            for (auto&& msg : r.errors())
+                System::print2("    ", msg, '\n');
+            Checks::exit_fail(VCPKG_LINE_INFO);
```
Comment (ras0219) [inline toolsrc/src/vcpkg/configuration.cpp:48]: We should not be printing nor terminating during any deserializers; this makes it painful to test and prevents composition.
```suggestion
            return Configuration{};
```

File: toolsrc/src/vcpkg/portfileprovider.cpp [hunk 2]
```diff
@@ -95,21 +86,25 @@ namespace vcpkg::PortFileProvider
         }
     }
 
-    static Optional<SourceControlFileLocation> try_load_overlay_port(const Files::Filesystem& fs,
-                                                                     View<fs::path> overlay_ports,
-                                                                     const std::string& spec)
+    static std::unique_ptr<OverlayRegistryEntry> try_load_overlay_port(const Files::Filesystem& fs,
```
Comment (ras0219) [inline toolsrc/src/vcpkg/portfileprovider.cpp:89]: This will need merging with https://github.com/microsoft/vcpkg/pull/15013. A key behavior of overlay ports in that PR is that they short-circuit versioning (similar to overrides). It appears that this PR turns overlays into effectively "simple" filesystem registries.

File: toolsrc/src/vcpkg/portfileprovider.cpp [hunk 3]
```diff
@@ -405,84 +347,65 @@ namespace vcpkg::PortFileProvider
...
-
         auto cache_it = m_impl->control_cache.find(version_spec);
         if (cache_it != m_impl->control_cache.end())
         {
             return cache_it->second;
         }
 
-        auto git_tree_cache_it = m_impl->git_tree_cache.find(version_spec);
-        if (git_tree_cache_it == m_impl->git_tree_cache.end())
+        auto entry_it = m_impl->entry_cache.find(version_spec.port_name);
```
Comment (ras0219) [inline toolsrc/src/vcpkg/portfileprovider.cpp:379]: `entry_it` appears only to be required if the `path_cache` misses -- this block should probably be moved inside the `path_cache` miss `if()` below.

File: toolsrc/src/vcpkg/registries.cpp [hunk 4]
```diff
@@ -16,349 +18,923 @@ namespace
...
+
+        View<VersionT> get_port_versions() const override { return port_versions; }
+        ExpectedS<fs::path> get_path_to_version(const VcpkgPaths&, const VersionT& version) const override;
 
-        BuiltinEntry(fs::path&& p) : port_directory(std::move(p)) { }
+        std::string port_name;
 
-        fs::path get_port_directory(const VcpkgPaths&, const VersionT&) const override { return port_directory; }
+        // these two map port versions to git trees
+        // these shall have the same size, and git_trees[i] shall be the git tree for port_versions[i]
```
Comment (ras0219) [inline toolsrc/src/vcpkg/registries.cpp:33]: sorted?

File: toolsrc/src/vcpkg/registries.cpp [hunk 5]
```diff
@@ -16,349 +18,923 @@ namespace
...
+        fs::path m_path;
+        std::string m_baseline_identifier;
+        DelayedInit<Baseline> m_baseline;
+    };
 
-            return std::pair<VersionT, fs::path>{std::move(version).value_or(VersionT{}), std::move(registry_path)};
+    ExpectedS<fs::path> get_git_baseline_json_path(const VcpkgPaths& paths, StringView baseline_commit_sha)
+    {
+        auto baseline_path = paths.git_checkout_baseline(paths.get_filesystem(), baseline_commit_sha);
+        if (paths.get_filesystem().exists(baseline_path))
```
Comment (ras0219) [inline toolsrc/src/vcpkg/registries.cpp:193]: This should be a post condition of `git_checkout_baseline()`. It should probably should return an `Optional` or `ExpectedS` to signal failure.

**Review by ras0219 — CHANGES_REQUESTED** [review]
(only reviewed to registries.cpp:193)

This will need a lot of tests added before merging; it should probably pull #15081 to make that easier. I suspect it will be faster to review and merge this PR if it was split into two:
1. Refactor existing Builtin+Filesystem registry into this new architecture (+tests for those)
2. Adding Git registries + tests

(Remember that code review costs are super-linear with number of lines changed!)

File: toolsrc/src/vcpkg/registries.cpp (hunk 5 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/registries.cpp:193]: This doesn't change any code, so I'm loath to change it.

**Comment by strega-nil** [comment]
This PR depends on #15114

File: toolsrc/src/vcpkg/registries.cpp (hunk 4 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/registries.cpp:33]: There is not necessarily a sorting here; in practice, it'll be the same as it is in the file.

File: toolsrc/src/vcpkg/portfileprovider.cpp (hunk 2 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/portfileprovider.cpp:89]: mhh, kind of. Technically, I don't believe that (in this PR anyways), overlay ports are even considered. This is just a "hack" to give overlay ports an entry when versioning isn't a factor.

File: toolsrc/include/vcpkg/versiondeserializers.h [hunk 6]
```diff
@@ -39,12 +32,4 @@ namespace vcpkg
...
                                    int port_version,
                                    bool always_emit_port_version = false);
-
-    ExpectedS<std::map<std::string, VersionT, std::less<>>> parse_baseline_file(Files::Filesystem& fs,
-                                                                                StringView baseline_name,
-                                                                                const fs::path& baseline_file_path);
-
-    ExpectedS<std::vector<VersionDbEntry>> parse_versions_file(Files::Filesystem& fs,
-                                                               StringView port_name,
-                                                               const fs::path& versions_file_path);
```
Comment (vicroms) [inline toolsrc/include/vcpkg/versiondeserializers.h:49]: These are currently used in #14999 to validate the integrity of the versioning files in the built-in registry.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 7]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
...
+                                        init_output.output),
+                        expected_right_tag};
+            }
+        }
+
+        auto lock_file = work_tree / fs::u8path(".vcpkg-lock");
+
+        std::error_code ec;
+        struct Guard
+        {
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:712]: IMO this should be a real thing, not a one-off local type.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 8]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
...
+                git_cmd_builder(*this, dot_git_dir, work_tree).string_arg("init");
+            auto init_output = System::cmd_execute_and_capture_output(init_registries_git_dir.extract());
+            if (init_output.exit_code != 0)
+            {
+                return {Strings::format("Error: Failed to initialize local repository %s.\n%s\n",
+                                        fs::u8string(work_tree),
+                                        init_output.output),
+                        expected_right_tag};
+            }
+        }
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:706]: Can you comment why there's an extra scope here?

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 9]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
...
+        if (get_filesystem().exists(git_tree_final))
+        {
+            return std::move(git_tree_final);
+        }
+
+#if defined(_WIN32)
+        long pid = ::_getpid();
+#else
+        long pid = ::getpid();
+#endif
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:802]: Seems like this should be in base?

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 10]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
...
+        auto git_rev_parse_output = System::cmd_execute_and_capture_output(git_rev_parse.extract());
+        if (git_rev_parse_output.exit_code != 0)
+        {
+            return {git_rev_parse_output.output, expected_right_tag};
+        }
+        return {Strings::trim(git_rev_parse_output.output).to_string(), expected_left_tag};
+    }
+    ExpectedS<fs::path> VcpkgPaths::git_checkout_object_from_remote_registry(StringView object) const
+    {
+        get_filesystem().create_directories(m_pimpl->registries_git_trees, VCPKG_LINE_INFO);
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:790]: Would prefer a field access over `get_filesystem` due to debugging pain (since step into steps into get_filesystem etc.)

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 11]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
         return destination;
     }
 
+    ExpectedS<std::string> VcpkgPaths::git_fetch_from_remote_registry(StringView repo, StringView treeish) const
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:690]: IMO this is making `VcpkgPaths` violate SRP: it's no longer a collection of paths, now it's paths and other stuff that vaguely touches the filesystem.

(But sadly I see this sin was already committed :( )

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 12]
```diff
@@ -679,6 +687,178 @@ If you wish to silence this error and use classic mode, you can:
...
+        }
+
+        std::error_code ec;
+        get_filesystem().rename(git_tree_temp, git_tree_final, ec);
+
+        if (get_filesystem().exists(git_tree_final))
+        {
+            return git_tree_final;
+        }
+        else if (ec)
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:852]: Drop else due to preceding return x2

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 11 above)
Comment (vicroms) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:690]: I think is not too late to refactor all of this out of `VcpkgPaths` and make them free functions.

Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:690]: I think it'll be pretty easy to do that whenever we want.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 9 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:802]: why? this is just creating a temporary file name.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 11 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:691]: @vicroms It's "easy" but it's also not something I would hold this PR over.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp (hunk 9 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:803]: Because it's doing a _WIN32 check. I think anything that is in "expose something from the system with platform selection" like this makes sense to be extracted into something like `vcpkg::get_pid`

Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:803]: hm, guess so.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 13]
```diff
@@ -679,6 +688,178 @@ If you wish to silence this error and use classic mode, you can:
...
+            .string_arg("-E")
+            .string_arg("tar")
+            .string_arg("xf")
+            .path_arg(git_tree_temp_tar);
+
+        auto untar_cmdline = untar.extract();
+#ifdef WIN32
+        // Invoke through `cmd` to support `&&`
+        untar_cmdline.insert(0, "cmd /c \"");
+        untar_cmdline.push_back('"');
```
Comment (vicroms) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:837]: This doesn't seem to be working on my end, no files are extracted.  
I believe this may be creating a badly escaped sequence.

```powershell
cmd /c "cd "%LOCALAPPDATA%\registries\git-trees\9558247e56da3b4be8a340982a166830d96b5ffd.tmp12284" &&
  "D:\vcpkg\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe" -E tar xf 
  "%LOCALAPPDATA%\registries\git-trees\9558247e56da3b4be8a340982a166830d96b5ffd.tmp12284.tar""
```

**Review by vicroms — DISMISSED** [review]

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 14]
```diff
@@ -202,6 +203,10 @@ namespace vcpkg
                 , m_env_cache(ff_settings.compiler_tracking)
                 , m_ff_settings(ff_settings)
             {
+                const auto& cache_root = System::get_platform_cache_home().value_or_exit(VCPKG_LINE_INFO);
+                registries_work_tree_dir = cache_root / fs::u8path("registries") / fs::u8path("git");
+                registries_dot_git_dir = registries_work_tree_dir / fs::u8path(".git");
+                registries_git_trees = cache_root / fs::u8path("registries") / fs::u8path("git-trees");
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:209]: oops...

**Review by strega-nil — COMMENTED** [review]
I didn't notice anything else in a cursory glance.

File: toolsrc/src/vcpkg/commands.porthistory.cpp [hunk 15]
```diff
@@ -27,24 +27,27 @@ namespace vcpkg::Commands::PortHistory
...
             System::CmdLineBuilder builder;
             builder.path_arg(git_exe)
                 .string_arg(Strings::concat("--git-dir=", fs::u8string(dot_git_directory)))
                 .string_arg(Strings::concat("--work-tree=", fs::u8string(working_directory)));
-            const std::string full_cmd = Strings::concat(builder.extract(), " ", cmd);
 
-            const auto output = System::cmd_execute_and_capture_output(full_cmd);
+            auto full_cmd = std::move(builder).extract();
+            full_cmd.append(" ");
+            full_cmd.append(cmd.begin(), cmd.end());
```
Comment (ras0219) [inline toolsrc/src/vcpkg/commands.porthistory.cpp:44]: `Strings::concat()` is optimized to reuse an rvalue string as the first parameter, so there's no need to write this out long-form.

```suggestion
            const auto full_cmd = Strings::concat(std::move(builder).extract(), " ", cmd);
```

File: toolsrc/include/vcpkg/base/system.process.h [hunk 16]
```diff
@@ -24,19 +24,39 @@ namespace vcpkg::System
...
+    {
+        const fs::path& working_directory;
+    };
+
     struct CmdLineBuilder
     {
-        CmdLineBuilder& path_arg(const fs::path& p) { return string_arg(fs::u8string(p)); }
-        CmdLineBuilder& string_arg(StringView s);
-        CmdLineBuilder& ampersand()
+        CmdLineBuilder() = default;
```
Comment (ras0219) [inline toolsrc/include/vcpkg/base/system.process.h:34]: It would be better if this refactor was a separate PR.

File: toolsrc/src/vcpkg/registries.cpp [hunk 17]
```diff
@@ -407,6 +477,118 @@ namespace
     }
     // } FilesystemRegistry::RegistryImplementation
 
+    // { GitRegistry::RegistryImplementation
+    std::unique_ptr<RegistryEntry> GitRegistry::get_port_entry(const VcpkgPaths& paths, StringView port_name) const
+    {
+        auto port_versions = get_versions_tree_path(paths);
```
Comment (ras0219) [inline toolsrc/src/vcpkg/registries.cpp:483]: Running `git fetch` for every registry on every `vcpkg install` is an unacceptably bad perf problem; this needs optimization work to avoid hitting the network on an otherwise no-op command (such as everything is already installed and cached locally).

Comment (ras0219) [reply toolsrc/src/vcpkg/registries.cpp:483]: A possible solution to this would be to require users to list the commit to use version info from in their manifest/configuration until a separate lock file is implemented.

**Review by ras0219 — CHANGES_REQUESTED** [review]
This also needs several end-to-end tests. I'd suggest having the end-to-end tests construct a local git repo with an appropriate history and then point to that temporary repo for testing.

**Comment by strega-nil** [comment]
> This also needs several end-to-end tests. I'd suggest having the end-to-end tests construct a local git repo with an appropriate history and then point to that temporary repo for testing.

It does, though? https://github.com/microsoft/vcpkg/pull/15054/files#diff-de6d1117ad4379767b02529167248f18bbad1b56758118cccfaa915d534be7b7R56

File: toolsrc/src/vcpkg/registries.cpp (hunk 17 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/registries.cpp:483]: I would argue that the vcpkg lock file is a necessity for stabilization.

**Comment by ras0219** [comment]
Oops, I managed to forget about the e2e tests by the time I got to the end of the diff 👍

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 18]
```diff
@@ -1,13 +1,164 @@
...
+    New-Item -Path 'vcpkg.json' -ItemType File -Value "
+{
+    `"name`": `"manifest-test`",
+    `"version-string`": `"1.0.0`",
+    `"dependencies`": [
+        `"vcpkg-internal-e2e-test-port`"
+    ]
+}
+"
+    New-Item -Path 'vcpkg-configuration.json' -ItemType File -Value "
```
Comment (ras0219) [inline scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:104]: This would be more correct with `ConvertTo-Json` instead of munging `$filesystemRegistry`:

```powershell
(@{ "default-registry"=$null; registries=@("a", "b") } | ConvertTo-Json -Depth 5)
```

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 19]
```diff
@@ -1,13 +1,164 @@
...
 
-$commonArgs += @("--x-builtin-port-versions-dir=$PSScriptRoot/../../e2e_ports/port_versions")
+$builtinRegistryArgs = $commonArgs + @("--x-builtin-port-versions-dir=$PSScriptRoot/../../e2e_ports/port_versions")
 
-Run-Vcpkg install @commonArgs 'vcpkg-internal-e2e-test-port'
+Run-Vcpkg install @builtinRegistryArgs 'vcpkg-internal-e2e-test-port'
 Throw-IfNotFailed
 
-Run-Vcpkg install @commonArgs --feature-flags=registries 'vcpkg-internal-e2e-test-port'
+Run-Vcpkg install @builtinRegistryArgs --feature-flags=registries 'vcpkg-internal-e2e-test-port'
```
Comment (ras0219) [inline scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:9]: I know this diff is not changing the status quo but this behavior is surprising to me.

Specifically, if I have edits out in `ports/vcpkg-internal-e2e-test-port`, turning on the registries feature flag should _not_ cause `vcpkg install vcpkg-internal-e2e-test-port` to use files found via `port_versions` -- that would be a huge workflow break. This isn't what's under test here, but it is a possible implication of what's under test (turning on the feature flag uses `port_versions` in some circumstances).

I'd really like to see a test that ensures continuity of experience for existing workflows such as editing the built-in ports. I suspect the easiest way to implement this test would be to have a version via `port_versions` fail to build.

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 20]
```diff
@@ -1,13 +1,164 @@
...
 Throw-IfFailed
 
-Run-Vcpkg install @commonArgs --feature-flags=registries 'zlib'
+Run-Vcpkg install @builtinRegistryArgs --feature-flags=registries 'zlib'
 Throw-IfFailed
+
+Run-Vcpkg x-set-installed @builtinRegistryArgs --feature-flags=registries
+Throw-IfFailed
```
Comment (ras0219) [inline scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:16]: If the intention of this is to "reset" everything to a pristine state and not carry over any `versioning`/`buildtrees`/etc files:
```suggestion
Refresh-TestRoot
```

File: scripts/buildsystems/vcpkg.cmake [hunk 21]
```diff
@@ -384,6 +384,13 @@ if(VCPKG_MANIFEST_MODE AND VCPKG_MANIFEST_INSTALL AND NOT _CMAKE_IN_TRY_COMPILE
             endforeach()
         endif()
 
+        if(DEFINED VCPKG_FEATURE_FLAGS OR DEFINED CACHE{VCPKG_FEATURE_FLAGS})
+            list(JOIN VCPKG_FEATURE_FLAGS "," _VCPKG_FEATURE_FLAGS)
+            set(_VCPKG_FEATURE_FLAGS "--feature-flags=${_VCPKG_FEATURE_FLAGS}")
+        elseif(DEFINED ENV{VCPKG_FEATURE_FLAGS})
+            set(_VCPKG_FEATURE_FLAGS "--feature-flags=$ENV{VCPKG_FEATURE_FLAGS}")
```
Comment (ras0219) [inline scripts/buildsystems/vcpkg.cmake:391]: Why is this needed?

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 (hunk 19 above)
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:9]: This does not make sense to me. What is your expected behavior, and what is the behavior right now?

Comment (BillyONeal) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:9]: The status quo is that work on the versioning feature and work on the registries feature both make edits to `ports/vcpkg-internal-e2e-test-port` in a way that might pass their PR's testing, but which upon merge are likely broken (e.g. because anything edited for the sake of testing git registries is not reflected in the version database).

I think Robert is asking those to be segregated somehow so that this phantom merge conflict becomes unlikely to happen.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 22]
```diff
@@ -724,6 +734,167 @@ If you wish to silence this error and use classic mode, you can:
...
+                                    fs::u8string(work_tree),
+                                    init_output.output),
+                    expected_right_tag};
+        }
+
+        auto lock_file = work_tree / fs::u8path(".vcpkg-lock");
+
+        std::error_code ec;
+        auto guard = make_unique_resource(
+            fs.take_exclusive_file_lock(lock_file, ec), fs::SystemHandle{}, [&fs](fs::SystemHandle h) {
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:760]: This doesn't check `ec` after `take_exclusive_file_lock`.

Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:760]: To follow up on my UniqueResource comment above, if this were like the following instead, you get an independently testable subcomponent and this bug is much clearer, because the work to acquire and release the resource are together and symmetric:

```
class ExclusiveFileLock {
    filesystem& fs_;
    fs::path path_;
    const char* entity_;
    SystemHandle handle_;

public:
    ExclusiveFileLock(filesystem& fs, fs::path path, const char* entity)
        : fs_(fs)
        , path_(path)
        , entity_(entity)
        , handle_() {
        error_code ec;
        handle_ = fs_.take_exclusive_file_lock(path_, ec);
        if (ec) {
            Debug::Print("failed to take lock of ", entity_, " (", path_, "));
            Debug::Print(ec.message());
        }
    }

    ~ExclusiveFileLock() {
        if (!handle_) { return; }
        std::error_code ec;
        fs.unlock_file_lock(handle_, ec);
        if (ec)
        {
            Debug::print("Warning: failed to drop file lock on ", entity_, ": ", ec.message());
        }
    }

    ExclusiveFileLock(const ExclusiveFileLock&) = delete;
    ExclusiveFileLock& operator=(const ExclusiveFileLock&) = delete;
};

// ... 

ExclusiveFileLock registry_lock{fs, lock_file, "git registry"};
```

Comment (BillyONeal) [reply toolsrc/src/vcpkg/vcpkgpaths.cpp:760]: (Even better might be to make the actual contents of `take_exclusive_file_lock` be in the ctor but I digress :) )

File: toolsrc/src/vcpkg/registries.cpp [hunk 23]
```diff
@@ -37,6 +40,73 @@ namespace
...
+                return std::move(*maybe_hash.get());
+            });
+        }
+
+        fs::path get_versions_tree_path(const VcpkgPaths& paths) const
+        {
+            return m_versions_tree.get([this, &paths]() -> fs::path {
+                auto maybe_tree = paths.git_find_object_id_for_remote_registry_path(get_commit_of_repo(paths),
+                                                                                    fs::u8path("port_versions"));
+                if (auto tree = maybe_tree.get())
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/registries.cpp:76]: I think this would be better if it followed the same pattern as `get_commit_of_repo` e.g.:

```
auto tree = maybe_tree.get();
if (!tree) { Checks::exit_with_message("Error: could not find the git tree" ....); }
auto maybe_path = paths.git_checkout_object_from_remote_registry(*tree);
if (!maybe_path) { Checks::exit_with_message("Error: could not find the git tree for `port_versions` ...); }
return std::move(*maybe_path.get());
```

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 24]
```diff
@@ -1,13 +1,163 @@
...
+New-Item -Path $filesystemRegistry -ItemType Directory
+$filesystemRegistry = (Get-Item $filesystemRegistry).FullName
+
+Copy-Item `
+    -LiteralPath "$PSScriptRoot/../../e2e_ports/port_versions/baseline.json" `
+    -Destination "$filesystemRegistry/baseline.json"
+Copy-Item -Recurse `
+    -LiteralPath "$PSScriptRoot/../../e2e_ports/vcpkg-internal-e2e-test-port" `
+    -Destination "$filesystemRegistry"
+New-Item `
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:30]: `mkdir` has different behavior depending on operating system, so I prefer to use built-in Powershell stuff.

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 25]
```diff
@@ -1,13 +1,163 @@
...
+    Throw-IfFailed
+    git config --local user.email 'nobody@example.com'
+    Throw-IfFailed
+    git config --local core.autocrlf 'false'
+    Throw-IfFailed
+    Copy-Item -Recurse -LiteralPath "$PSScriptRoot/../../e2e_ports/port_versions" -Destination .
+    Copy-Item -Recurse -LiteralPath "$PSScriptRoot/../../e2e_ports/vcpkg-internal-e2e-test-port" -Destination .
+
+    $CurrentTest = 'git add .'
+    git add .
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:72]: yeah probably, TIL about `-A` 😝

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 26]
```diff
@@ -1,13 +1,163 @@
...
+# test the filesystem registry
+$manifestDir = "$TestingRoot/filesystem-registry-test-manifest-dir"
+
+New-Item -Path $manifestDir -ItemType Directory
+$manifestDir = (Get-Item $manifestDir).FullName
+
+Push-Location $manifestDir
+try
+{
+    $vcpkgJson = @{
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:94]: I think this is better, but I'm not super attached.

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 [hunk 27]
```diff
@@ -1,13 +1,163 @@
...
+    }
+    New-Item -Path 'vcpkg.json' -ItemType File `
+        -Value (ConvertTo-Json -Depth 5 -InputObject $vcpkgJson)
+
+    $vcpkgConfigurationJson = @{
+        "default-registry" = $null;
+        "registries" = @(
+            @{
+                "kind" = "filesystem";
+                "path" = $filesystemRegistry;
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:109]: it does not.

File: scripts/azure-pipelines/end-to-end-tests-prelude.ps1 [hunk 28]
```diff
@@ -1,3 +1,5 @@
+$WorkingRoot = (Get-Item $WorkingRoot).FullName
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-prelude.ps1:1]: that's a fair point.

File: scripts/buildsystems/vcpkg.cmake [hunk 29]
```diff
@@ -384,6 +384,11 @@ if(VCPKG_MANIFEST_MODE AND VCPKG_MANIFEST_INSTALL AND NOT _CMAKE_IN_TRY_COMPILE
             endforeach()
         endif()
 
+        if(DEFINED VCPKG_FEATURE_FLAGS OR DEFINED CACHE{VCPKG_FEATURE_FLAGS})
+            list(JOIN VCPKG_FEATURE_FLAGS "," _VCPKG_FEATURE_FLAGS)
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply scripts/buildsystems/vcpkg.cmake:388]: this is a CMake argument, not an environment variable; it should be possible to, say, write:

```
set(VCPKG_FEATURE_FLAGS manifests registries versioning)
```

File: scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1 (hunk 19 above)
Comment (strega-nil) [reply scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1:9]: I still feel like I don't understand what the issue is here.

File: toolsrc/include/vcpkg/base/unique_resource.h [hunk 30]
```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <type_traits>
+#include <utility>
+
+namespace vcpkg
+{
+    template<class T, class D>
+    struct UniqueResource
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply toolsrc/include/vcpkg/base/unique_resource.h:9]: That was originally what this PR did. I made the change because it was requested. I personally prefer your method, and I will switch it back, but I believe @ras0219 complained last time.

File: toolsrc/src/vcpkg/vcpkgpaths.cpp [hunk 31]
```diff
@@ -724,6 +734,167 @@ If you wish to silence this error and use classic mode, you can:
...
+        }
+
+        std::error_code ec;
+        fs.rename(git_tree_temp, git_tree_final, ec);
+
+        if (fs.exists(git_tree_final))
+        {
+            return git_tree_final;
+        }
+        if (ec)
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/vcpkgpaths.cpp:886]: This ec check needs to happen before the exists. And then the exists probably needs to use ec.

**Review by ras0219 — APPROVED** [review]
